backend/compact-connect/common_constructs/user_pool.py

```diff
@@ -117,19 +117,53 @@ def __init__(  # pylint: disable=too-many-arguments
 
     def add_ui_client(
         self,
-        callback_urls: list[str],
+        ui_domain_name: str,
+        environment_context: dict,
         read_attributes: ClientAttributes,
         write_attributes: ClientAttributes,
         ui_scopes: list[OAuthScope] = None,
     ):
         """
         Creates an app client for the UI to authenticate with the user pool.
 
-        :param callback_urls: The URLs that Cognito allows the UI to redirect to after authentication.
+        :param ui_domain_name: The ui domain name used to determine acceptable redirects.
+        :param environment_context: The environment context used to determine acceptable redirects.
         :param read_attributes: The attributes that the UI can read.
         :param write_attributes: The attributes that the UI can write.
         :param ui_scopes: OAuth scopes that are allowed with this client
         """
+        callback_urls = []
+        if ui_domain_name is not None:
+            callback_urls.append(f'https://{ui_domain_name}/auth/callback')
+        # This toggle will allow front-end devs to point their local UI at this environment's user pool to support
+        # authenticated actions.
+        if environment_context.get('allow_local_ui', False):
+            local_ui_port = environment_context.get('local_ui_port', '3018')
+            callback_urls.append(f'http://localhost:{local_ui_port}/auth/callback')
+        if not callback_urls:
+            raise ValueError(
+                "This app requires a callback url for its authentication path. Either provide 'domain_name' or set "
+                "'allow_local_ui' to true in this environment's context."
+            )
+
+        logout_urls = []
+        if ui_domain_name is not None:
+            logout_urls.append(f'https://{ui_domain_name}/Login')
+            logout_urls.append(f'https://{ui_domain_name}/Logout')
+        # This toggle will allow front-end devs to point their local UI at this environment's user pool to support
+        # authenticated actions.
+        if environment_context.get('allow_local_ui', False):
+            local_ui_port = environment_context.get('local_ui_port', '3018')
+            logout_urls.append(f'http://localhost:{local_ui_port}/Login')
+            logout_urls.append(f'http://localhost:{local_ui_port}/Logout')
+        if not logout_urls:
+            raise ValueError(
+                "This app requires a logout url for its logout function. Either provide 'domain_name' or set "
+                "'allow_local_ui' to true in this environment's context."
+            )
+        # Do not allow resource server scopes via the client - they are assigned via token customization
+        # to allow for user attribute-based access
```

Comment:
Trying to parse this comment. What code does it apply to?

Reply:
I think this comment was erroneously copy pasted